### PR TITLE
[CDAP-12228] Rules engine End-to-End Integration

### PR DESCRIPTION
**Pull request by ajainarayanan**
- Adds the ability to,
    - import
    - delete
    - add to pipeline
- Adds rules engine directive in angular to open up rules engine in angular
- Fixes styling for angular-react compatibility
- Fixes server.js to pass on the headers while uploading file from client.

Related PR in yare plugin: https://github.com/cask-solutions/yare/pull/1

### CODE REVIEW

File: cdap-ui/app/cdap/components/RulesEngineHome/RuleBookDetails/RulebookMenu/AddRulesEngineToPipelineModal.js [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+    realtimeUrl: '',
+    batchConfig: null,
+    realtimeConfig: null
+  };
+
+  updateBatchAndRealtimeConfigs = () => {
+    let {selectedNamespace: namespace} = NamespaceStore.getState();
+    MyArtifactApi
+      .list({namespace})
+      .combineLatest([
```
Comment (elfenheart) [inline cdap-ui/app/cdap/components/RulesEngineHome/RuleBookDetails/RulebookMenu/AddRulesEngineToPipelineModal.js]: doesn't need to be in an array if just combining with 1

File: cdap-ui/app/cdap/components/RulesEngineHome/RuleBookDetails/RulebookMenu/AddRulesEngineToPipelineModal.js [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+      .subscribe(
+        (res) => {
+          let artifacts = res[0];
+          let rulebook = res[1].values[0];
+          let batchArtifacts = artifacts.filter(artifact => artifact.name === 'cdap-data-pipeline');
+          let realtimeArtifacts = artifacts.filter(artifact => artifact.name === 'cdap-data-streams');
+          let yareArtifact = artifacts.filter(artifact => artifact.name === 'yare-plugins');
+          let batchHighestArtifact, realtimeHighestArtifact;
+          if (batchArtifacts.length > 1) {
+            batchHighestArtifact = findHighestVersion(batchArtifacts.map(artifact => artifact.version), true);
```
Comment (elfenheart) [inline cdap-ui/app/cdap/components/RulesEngineHome/RuleBookDetails/RulebookMenu/AddRulesEngineToPipelineModal.js]: I don't think this returns the highest artifact object, i think this just returns the highest version string or version object

File: cdap-ui/app/cdap/components/RulesEngineHome/RuleBookDetails/RulebookMenu/RulebookMenu.scss [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+    .dropdown-menu {
+      .dropdown-item {
+        padding-left: 10px;
+        padding-right: 10px;
+
+        span {
+          margin: 0 10px;
+        }
+        hr {
+          margin-left: -10px;
```
Comment (elfenheart) [inline cdap-ui/app/cdap/components/RulesEngineHome/RuleBookDetails/RulebookMenu/RulebookMenu.scss]: `margin: 0 -10px`

File: cdap-ui/app/cdap/components/RulesEngineHome/ImportRulebookWizard/RulebookUploadStep.js [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+
+const mapStateToFileDnDProps = (state) => {
+  return {
+    file: state.upload.file
+  };
+};
+const mapDispatchToDndProps = (dispatch) => {
+  return {
+    onDropHandler: (e) => {
+      if (e[0]) {
```
Comment (tonybach) [inline cdap-ui/app/cdap/components/RulesEngineHome/ImportRulebookWizard/RulebookUploadStep.js]: nit: Maybe `if (!e[0]) { return; }` to have a smaller if condition?

File: cdap-ui/app/cdap/components/RulesEngineHome/ImportRulebookWizard/index.js [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+import ImportRulebookStore from 'components/RulesEngineHome/ImportRulebookWizard/ImportRulebookStore';
+import T from 'i18n-react';
+
+require('./ImportRulebookWizard.scss');
+const PREFIX = 'features.RulesEngine.ImportRulebook';
+
+export default class ImportRulebookWizard extends Component {
+
+  static propTypes = {
+    isOpen: PropTypes.book,
```
Comment (tonybach) [inline cdap-ui/app/cdap/components/RulesEngineHome/ImportRulebookWizard/index.js]: typo, should be `bool`

File: cdap-ui/app/cdap/components/RulesEngineHome/ImportRulebookWizard/index.js [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+  state = {
+    isOpen: this.props.isOpen || false,
+    file: ''
+  };
+
+  setFile = ([file]) => {
+    this.setState({
+      file
+    });
+  }
```
Comment (tonybach) [inline cdap-ui/app/cdap/components/RulesEngineHome/ImportRulebookWizard/index.js]: nit: add semi-colons at the end of these fat arrow functions

File: cdap-ui/app/cdap/components/RulesEngineHome/ImportRulebookWizard/index.js [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+
+  toggleWizard = () => {
+    this.setState({
+      isOpen: !this.state.showWizard
+    });
+    if (this.props.onClose) {
```
Comment (tonybach) [inline cdap-ui/app/cdap/components/RulesEngineHome/ImportRulebookWizard/index.js]: so do we call this even `this.state.isOpen` is true?

File: cdap-ui/app/cdap/components/RulesEngineHome/ImportRulebookWizard/index.js [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+        title={T.translate(`${PREFIX}.shorttitle`)}
+        isOpen={this.state.isOpen}
+        toggle={this.toggleWizard.bind(this, false)}
+        className="rulebook-upload-wizard"
+      >
+        <Wizard
+          wizardConfig={ImportRulebookWizardConfig}
+          store={ImportRulebookStore}
+          onSubmit={this.onSubmit.bind(this)}
+          onClose={this.toggleWizard.bind(this)
```
Comment (tonybach) [inline cdap-ui/app/cdap/components/RulesEngineHome/ImportRulebookWizard/index.js]: do we still need these bindings?

File: cdap-ui/app/cdap/components/RulesEngineHome/RuleBookDetails/RulebookMenu/AddRulesEngineToPipelineModal.js [hunk 9]
```diff
@@ -0,0 +1,192 @@
...
+          <div className="message">
+            {T.translate(`${PREFIX}.message`)}
+          </div>
+          <div className="action-buttons">
+            <a
+              className="btn btn-secondary"
+              href={this.state.batchUrl}
+              onClick={this.handleOnBatchUrlClick}
+            >
+              <i className="fa icon-ETLBatch"/>
```
Comment (tonybach) [inline cdap-ui/app/cdap/components/RulesEngineHome/RuleBookDetails/RulebookMenu/AddRulesEngineToPipelineModal.js]: nit: add space before `/>`

File: cdap-ui/app/cdap/components/RulesEngineHome/RuleBookDetails/RulebookMenu/AddRulesEngineToPipelineModal.js [hunk 10]
```diff
@@ -0,0 +1,192 @@
...
+              <span>{T.translate(`${PREFIX}.batchPipelineBtn`)}</span>
+            </a>
+            <a
+              href={this.state.realtimeUrl}
+              className={classnames('btn btn-secondary', {
+                'inactive': !this.state.realtimeUrl
+              })}
+              onClick={this.handleOnRealtimeUrlClick}
+            >
+              <i className="fa icon-sparkstreaming"/>
```
Comment (tonybach) [inline cdap-ui/app/cdap/components/RulesEngineHome/RuleBookDetails/RulebookMenu/AddRulesEngineToPipelineModal.js]: here too

File: cdap-ui/app/cdap/components/RulesEngineHome/RuleBookDetails/index.js [hunk 11]
```diff
@@ -147,31 +145,53 @@ export default class RuleBookDetails extends Component {
...
+              integration.embedded ?
+                <button
+                  className="btn btn-primary"
+                  onClick={this.onApply}
+                  disabled={this.state.onApplying}
+                >
+                  {
+                    this.state.onApplying ? <LoadingSVG /> : null
+                  }
+                  <span>Apply</span>
```
Comment (tonybach) [inline cdap-ui/app/cdap/components/RulesEngineHome/RuleBookDetails/index.js]: nit: i18n

**Comment by tonybach** [comment]
Code LGTM 👍

**Comment by leacd** [comment]
LGTM

**Comment by ajainarayanan** [comment]
Bamboo build: https://builds.cask.co/browse/CDAP-DRC7093-7

**Pull request merged**